bindings/sundials4py/arkode/arkode_usersupplied.hpp

```diff
@@ -0,0 +1,609 @@
+  nb::object sprkstep_f1;
+  nb::object sprkstep_f2;
+
+  // lsrkstep-specific user-supplied function pointers
+  nb::object lsrkstep_f;
+  nb::object lsrkstep_domeig;
+
+  // mristep-specific user-supplied function pointers
+  nb::object mristep_fse;
+  nb::object mristep_fsi;
+  nb::object mristep_preinnerfn;
+  nb::object mristep_postinnerfn;
+};
+
+struct mristepinnerstepper_user_supplied_fn_table
+{
+  nb::object mristepinner_evolvefn;
+  nb::object mristepinner_fullrhsfn;
+  nb::object mristepinner_resetfn;
+  nb::object mristepinner_getaccumulatederrorfn;
+  nb::object mristepinner_resetaccumulatederrorfn;
+  nb::object mristepinner_setrtolfn;
+};
+
+///////////////////////////////////////////////////////////////////////////////
+// ARKODE user-supplied functions
+///////////////////////////////////////////////////////////////////////////////
+
+inline arkode_user_supplied_fn_table* arkode_user_supplied_fn_table_alloc()
+{
+  // We must use malloc since ARKodeFree calls free
+  auto fn_table = static_cast<arkode_user_supplied_fn_table*>(
+    std::malloc(sizeof(arkode_user_supplied_fn_table)));
+
+  // Zero out the memory
+  std::memset(fn_table, 0, sizeof(arkode_user_supplied_fn_table));
+
+  return fn_table;
+}
+
+inline arkode_user_supplied_fn_table* get_arkode_fn_table(void* ark_mem)
+{
+  auto mem      = static_cast<ARKodeMem>(ark_mem);
+  auto fn_table = static_cast<arkode_user_supplied_fn_table*>(mem->python);
+  if (!fn_table)
+  {
+    throw sundials4py::null_function_table(
+      "Failed to get Python function table from ARKODE memory");
+  }
+  return fn_table;
+}
+
+inline mristepinnerstepper_user_supplied_fn_table* mristepinnerstepper_user_supplied_fn_table_alloc()
+{
+  // We must use malloc since ARKodeFree calls free
+  auto fn_table = static_cast<mristepinnerstepper_user_supplied_fn_table*>(
+    std::malloc(sizeof(mristepinnerstepper_user_supplied_fn_table)));
+
+  // Zero out the memory
+  std::memset(fn_table, 0, sizeof(mristepinnerstepper_user_supplied_fn_table));
+
+  return fn_table;
+}
+
+using ARKRootStdFn = int(sunrealtype t, N_Vector y, sundials4py::Array1d gout,
+                         void* user_data);
+
+inline int arkode_rootfn_wrapper(sunrealtype t, N_Vector y,
+                                 sunrealtype* gout_1d, void* user_data)
+{
+  auto fn_table = get_arkode_fn_table(user_data);
+  auto fn       = nb::cast<std::function<ARKRootStdFn>>(fn_table->rootfn);
+  auto nrtfn    = static_cast<ARKodeMem>(user_data)->root_mem->nrtfn;
+
+  sundials4py::Array1d gout(gout_1d, {static_cast<unsigned long>(nrtfn)},
+                            nb::find(gout_1d));
+
+  int status = fn(t, y, gout, nullptr);
+
+  return status;
+}
+
+template<typename... Args>
+inline int arkode_ewtfn_wrapper(Args... args)
+{
+  return sundials4py::user_supplied_fn_caller<
+    std::remove_pointer_t<ARKEwtFn>, arkode_user_supplied_fn_table, ARKodeMem,
+    1>(&arkode_user_supplied_fn_table::ewtn, std::forward<Args>(args)...);
+}
```

Comment:
For perfect forwarding
```suggestion
inline int arkode_ewtfn_wrapper(Args&&... args)
{
  return sundials4py::user_supplied_fn_caller<
    std::remove_pointer_t<ARKEwtFn>, arkode_user_supplied_fn_table, ARKodeMem,
    1>(&arkode_user_supplied_fn_table::ewtn, std::forward<Args>(args)...);
}
```
Although if the arguments are just POD, then you could remove `std::forward` instead. I won't repeat this for the other functions below.